processor/src/tests/debug.rs

```diff
@@ -0,0 +1,365 @@
+    ├──     1: 2
+    ├──     2: 3
+    ├──     3: 0
+    ├──     4: 5
+    └──     5: 0
+    State of procedure local 4 before step 29: 5
+    State of procedure local 6 before step 29: EMPTY
+    State of procedure locals [2, 6] before step 29:
+    ├──     2: 3
+    ├──     3: 0
+    ├──     4: 5
+    ├──     5: 0
+    └──     6: EMPTY
+    Memory state before step 34 for the context 0:
+    ├── 0x40000000: 0
+    ├── 0x40000001: 0
+    ├── 0x40000002: 1
+    ├── 0x40000003: 2
+    ├── 0x40000004: 3
+    ├── 0x40000005: 0
+    ├── 0x40000006: 5
+    └── 0x40000007: 0
+    ");
+}
+
+#[test]
+fn test_debug_adv_stack() {
+    let stack_input = [0, 1, 2, 3, 4]; // 4 is on top
+    let advice_stack = [8, 7, 6, 5, 4, 3, 2, 1]; // 8 is on top
+    let source = "
+        begin
+            # => [4, 3, 2, 1, 0]
+            debug.adv_stack   # => [8..1]
+            debug.adv_stack.2 # => [8, 7]
+            debug.adv_stack.0 # => [8..1]
+
+            padw adv_loadw
+            # => [5, 6, 7, 8, 4, 3, 2, 1, 0]
+            debug.adv_stack # => [4, 3, 2, 1, 0]
+            debug.stack.9
+            push.[8, 7, 6, 5] assert_eqw
+            # => [4, 3, 2, 1, 0]
+
+            adv_push.1
+            # => [4, 4, 3, 2, 1, 0]
+            debug.stack.6
+            debug.adv_stack # => [3, 2, 1]
+
+            # Check that we popped 4
+            assert_eq
+            # => [3, 2, 1, 0]
+
+            # Pops the remaining elements one-by-one
+            adv_push.3
+            # => [1, 2, 3, 3, 2, 1, 0]
+            debug.stack.7
+
+            # Check
+            push.0 reversew assert_eqw
+
+            # advice stack is empty
+            debug.adv_stack
+        end";
+
+    // Execute with debug buffer
+    let test = build_debug_test!(source, &stack_input, &advice_stack);
+    let (_stack, output) = test.execute_with_debug_buffer().expect("execution failed");
+
+    insta::assert_snapshot!(output, @r"
+    Advice Stack state before step 1:
+    ├──  0: 8
+    ├──  1: 7
+    ├──  2: 6
+    ├──  3: 5
+    ├──  4: 4
+    ├──  5: 3
+    ├──  6: 2
+    └──  7: 1
+    Advice Stack state before step 1:
+    ├──  0: 8
+    └──  1: 7
+    └── (6 more items)
+    Advice Stack state before step 1:
+    ├──  0: 8
+    ├──  1: 7
+    ├──  2: 6
+    ├──  3: 5
+    ├──  4: 4
+    ├──  5: 3
+    ├──  6: 2
+    └──  7: 1
+    Advice Stack state before step 6:
+    ├──  0: 4
+    ├──  1: 3
+    ├──  2: 2
+    └──  3: 1
+    Stack state before step 6:
+    ├──  0: 5
+    ├──  1: 6
+    ├──  2: 7
+    ├──  3: 8
+    ├──  4: 4
+    ├──  5: 3
+    ├──  6: 2
+    ├──  7: 1
+    └──  8: 0
+    Stack state before step 23:
+    ├──  0: 4
+    ├──  1: 4
+    ├──  2: 3
+    ├──  3: 2
+    ├──  4: 1
+    └──  5: 0
+    Advice Stack state before step 23:
+    ├──  0: 3
+    ├──  1: 2
+    └──  2: 1
+    Stack state before step 28:
+    ├──  0: 1
+    ├──  1: 2
+    ├──  2: 3
+    ├──  3: 3
+    ├──  4: 2
+    ├──  5: 1
+    └──  6: 0
+    Advice Stack empty before step 45.
```

Comment:
Would be great to print something out here - otherwise, it may not be immediately clear that the advice stack is empty.